gen.c

```diff
@@ -43,10 +67,23 @@ gen_worker (void *context)
           return (NULL);
         }
 
-      task_t current_task = *gen_context->state.task;
+      task_t current_task;
+
+      switch (gen_context->config->brute_mode)
```

Comment:
This is not elegant enough.
Let's introduce base type for rec_state_t and iter_state_t. 
typedef struct base_state_t {
  task_t * task;
} base_state_t;

Derive both types from this base type - i.e. just have it as a first member of the structs. This also will allow to avoid untyped pointers in gen_context_t and ,ethod next.